CONTRIBUTING.md

```diff
@@ -0,0 +1,121 @@
+
+- use of `t` manifest types in modules, when it makes sense;
+- conversion functions respecting the naming scheme: `t_of_xxx` or `xxx_of_t`.
+  For example, when converting to or from data described by
+  [Yojson](https://ocaml.org/p/yojson/latest), use `t_of_yojson` and
+  `yojson_of_t`.
+
+Changes unrelated to the issue addressed by a PR should be made in a separate
+PR. Additionally, formatting changes in parts of the code not concerned by a
+specific PR should be proposed in another PR.
+
+Ideally, any opened issue should be accompanied by a test with a reproduction.
+When working on a fix for an issue, the first commit should contain the test
+showing the issue. Following commits should fix the issue and update the test
+result accordingly.
+
+### Repository organization
+
+The repository exposes a number of separate libraries (some of which are
+internal) and vendor libraries (to reduce the dependencies required by the
+project). Here is a list of the libraries exposed by the project.
+
+#### ocaml-lsp-server
+
+Contains the concrete implementation of a protocol server language for OCaml. A
+frontend used in particular by [Visual Studio
+Code](https://github.com/ocamllabs/vscode-ocaml-platform), but also by code
+editors supporting LSP. The code lives mainly in the following directories:
+[ocaml-lsp-server/](ocaml-lsp-server/).
+
+In addition, the project exposes two sub-directories dedicated to [code
+actions](ocaml-lsp-server/src/code_actions) and [custom
+requests](https://github.com/ocaml/ocaml-lsp/tree/master/ocaml-lsp-server/src/custom_requests).
+
+In most cases, it is likely that the contributions will focus solely on this
+project.
+
+##### Warning
+
+For historical reasons, but also for development convenience, `ocaml-lsp-server`
+should not build logic based on `Parsetree` (which changes from version to
+version and migration logic is not provided by `lsp`, nor `ocaml-lsp-server` but
+by [Merlin](https://github.com/ocaml/merlin)). If a command, or a constant,
+relies on the `Parsetree`, it is **strongly** recommended to build a command in
+Merlin that handles the logic to ensure the migration is localized to a single
+project.
+
+#### lsp
+
+Implementation of the LSP protocol in OCaml. It is designed to be as portable as
+possible and does not make any assumptions about IO. This is the implementation
+of the plumbing required to describe the LSP protocol and is used by a concrete
+server (for example the OCaml server) to describe the exposition of the
+protocol. The code lives mainly in the following directories: [lsp/](lsp/) and
+[lsp-fiber/](lsp-fiber/).
+
+##### Warning
+
+The set of types forming the LSP protocol API is generated automatically by a
+[preprocessor](lsp/bin) based on the [protocol
+specification](https://microsoft.github.io/language-server-protocol/overviews/lsp/overview/).
+The pair of [types.ml](lsp/src/types.ml) and [types.mli](lsp/src/types.mli)
+files must therefore never be modified manually.
```

Comment:
They can actually be modified manually. Only the parts inside the cinaps comment marks cannot be modified.